docs/proposals/network-policy.md

```diff
@@ -0,0 +1,300 @@
+  // is empty, then traffic should not be restricted based on port.  If this 
+  // field is not empty, traffic must match at least one NetworkPolicyPort in the list
+  // or else it will be dropped.
+  Ports []NetworkPolicyPort `json:"ports,omitempty"`
+
+  // List of sources which should be able to access the pods selected by PodSelector.
+  // Items in this list are combined using a logical OR operation.
+  // If this field is empty, then traffic should not be restricted based on 
+  // source.  If this field is not empty, traffic must match at least one 
+  // NetworkPolicySource in the list or else it will be dropped.
+  From []NetworkPolicySource `json:"from,omitempty"`
+}
+
+type NetworkPolicyPort struct {
+  // The protocol (TCP or UDP) which traffic must match.
+  // If not defined, this field defaults to TCP. 
+  Protocol api.Protocol `json:"protocol"`
+
+  // If specified, the port on the given protocol.  This can 
+  // either be a numerical or named port.  If not defined,
+  // this NetworkPolicyPort does not restrict based on port number.
+  // If defined, only traffic on the specified protocol AND port
+  // will be matched by this NetworkPolicyPort.
+  Port *intstr.IntOrString `json:"port,omitempty"`
+}
+
+type NetworkPolicySource struct {
+  // If 'Namespaces' is defined, 'Pods' must not be.
+  // This is a label selector which selects Pods in this namespace.
+  // This NetworkPolicySource matches any pods selected by this selector.
+  Pods *unversioned.LabelSelector `json:"pods,omitempty"`
+
+  // If 'Pods' is defined, 'Namespaces' must not be.
+  // Selects Kubernetes Namespaces.  This NetworkPolicySource matches 
+  // all pods in all namespaces selected by this label selector. 
+  Namespaces *unversioned.LabelSelector `json:"namespaces,omitempty"`
+}
+```
+
+### Behavior
+
+The following pseudo-code attempts to define when traffic is allowed to a given pod when using this API.
+
+```python
+def is_traffic_allowed(traffic, pod):
+  """
+  Returns True if traffic is allowed to this pod, False otherwise.
+  """
+  if not namespace.networkPolicy.ingress.isolation:
+    # If ingress isolation is disabled on the Namespace, all traffic is allowed.
+    return True 
+  elif traffic.source == pod.node:
+    # Traffic is from the pod's host - this allows for kubelet health checks.
+    return True
+  else:
+    # If namespace ingress isolation is enabled, only allow traffic 
+    # that matches a network policy which selects this pod.
+    for network_policy in network_policies(pod.namespace):
+      if not network_policy.podSelector.selects(pod):
+        # This policy doesn't select this pod. Try the next one. 
+        continue
+
+      # This policy selects this pod.  Check each ingress rule 
+      # defined on this policy to see if it allows the traffic.
+      for ingress_rule in network_policy.ingress:
+        if ingress_rule.from.matches(traffic) and \
```

Comment:
`ingress_rule.from` can be a null pointer.
This doc should define the `matches` method too.